### ISSUE TITLE
Convert should be able to convert types into any type with a single argument constructor

### ISSUE THREAD

**Issue by axelfontaine** [jOOQ/jOOQ#10229]
### Problem

When building a domain model, it is often very valuable to encode validation rules together with the data.

Say we have a simple POJO we map to a record (Using Kotlin syntax for conciseness):
```kotlin
data class Race(
  val id: Long,
  val distance: Int
)
```

We could validate input at the every edge of the system using
```kotlin
if (distance < 0 || distance > 1000) {
  throw IllegalArgumentException("Invalid distance: $distance")
}
val race = Race(id, distance)
```

Doing things like this unfortunately requires the developer to carefully guard every edge, as missing one means that invalid data could penetrate the system.

A better approach would be to integrate this logic directly into a Value Type (per DDD meaning):
```kotlin
data class RaceDistance(val d: Int) {
  init {
    require(distance > 0 || distance < 1000) { "Invalid distance: $distance" }
  }
}

data class Race(
  val id: Long,
  val distance: RaceDistance
)
```

This way it simple becomes impossible to create a race with an invalid distance and all edges of the system are automatically covered.

Unfortunately jOOQ currently doesn't know how to automatically map a record generated from a table like
```sql
CREATE TABLE race
(
    id            IDENTITY,
    distance      INT NOT NULL
);
```
onto this POJO as it required a forcedType and a custom converter to handle the `distance` field. While this works, it is tedious to implement.

### Solution

I would therefore like to propose a simple addition to `org.jooq.tools.Convert.ConvertAll.from(Object from)` that covers this case generically.

All that is needed is an additional fallback, in case all other attempts have failed to map the `from` parameter using `Reflect.onClass(toClass).create(from).get()` with the necessary exception handling in case the constructor doesn't exist. No other changes are required.

This alone would eliminate over 90% of all forcedType definitions I have ever written for jOOQ. More broadly this would encourage jOOQ users to build more robust domain models by removing a major pain point from the status quo.

**Comment by axelfontaine** [comment]
Looking at the current implementation, this could also simplify things quite a bit as it would eliminate the need for the following conversions, which would now be handled generically:

- `new URI(from.toString())`
- `new File(from.toString())`

as well as a good number of the various date conversions.

Extending this to also cover `of` and `valueOf` static methods would even cover a lot more cases.

**Comment by axelfontaine** [comment]
Or to put it differently, think of this as an attack against the stringly-typed world... ;-) 
https://twitter.com/lukaseder/status/1266019677148495872

**Comment by lukaseder** [comment]
Thanks for your suggestions. A better version of this feature is on its way under the name "embedded data type" (just like JPA's `@Embeddable`): https://github.com/jOOQ/jOOQ/issues/2530, https://github.com/jOOQ/jOOQ/issues/8353. The feature has a draft implementation, which is not documented yet, and might change.

To jOOQ, it doesn't matter if such an embedded type wraps a single column (your special case), or several columns (general case). The wrapping/unwrapping logic is the same, and so is the user experience.

I could imagine, however, a simpler setup in the code generator to specify the behaviour for single-column embeddable types. And yes, in Kotlin, those embeddable types should be `data classes`.

Other interesting, related feature requests:

- https://github.com/jOOQ/jOOQ/issues/8843 "Add new expression in <forcedType/> that matches check constraints" (because you surely have a `CHECK` constraint or even a `DOMAIN` modelling your validation also on the server side)
- https://github.com/jOOQ/jOOQ/issues/4543 "Add support for client-side DataType validation". This could even be extended to trying to reproduce the SQL `CHECK` expression in Java. For many expressions, a Java equivalent should be doable.

I'm saying "a better version", because ultimately, you will want a `Field<RaceDistance>` in your generated code for type safe `INSERT`, `UPDATE`, etc. as well as type safe comparisons `DISTANCE.eq(RaceDistance(1))`. The `Int` type should disappear from your client logic.

However, I agree with you there seems to be a low hanging fruit here if trying to convert a database type `T` to a "POJO" with a single-parameter constructor whose parameter is compatible with `T`

**Comment by lukaseder** [comment]
One issue here is that currently, there's no caching of reflection work in this `Convert` class. While looking up `object.getClass()` or `Class.isAssignableFrom()` calls are relatively fast, reflection on constructors is not.

We can solve that later, though.

**Comment by lukaseder** [comment]
There is also the open question of what to do with several applicable constructors. Since this logic is static, we cannot configure or override it. A reasonable default is to take the first declared single argument, applicable¹ constructor that is provided by `Class::getConstructors` (public), and if none was found, move on to `Class::getDeclaredConstructors` (public and private). If there are several applicable constructors, then the behaviour is undefined as the Java reflection API does not provide any guarantees with respect to ordering of these constructors.

¹: Applicable here means that the source value (e.g. of type `String`) can be converted to the constructor parameter type (e.g. of type `int`)

**Comment by lukaseder** [comment]
There's another caveat in this feature. There will be a risk of not being forward compatible in some cases. E.g. take `ArrayList`. With the new feature, converting an `int` to an `ArrayList` will call the `ArrayList(int)` constructor, where `int` is a capacity. We might want to support converting the single `int` value into an `ArrayList<Integer>` where the list contains that single `int` value. Currently, we could implement this backwards compatibly. This would no longer be the case.

**Comment by axelfontaine** [comment]
What would be the use-case for ever wanting to convert an int db column into a Java ArrayList? Just trying to understand where you're coming from...

**Comment by lukaseder** [comment]
> What would be the use-case for ever wanting to convert an int db column into a Java ArrayList? Just trying to understand where you're coming from...

1) The first random example I could come up with. 2) If someone can see that this works, someone will find a use-case.

**Comment by lukaseder** [comment]
Anyway, I really like the idea. Out of the box, we can now map into all sorts of types, including e.g. things like `long -> AtomicLong`, because, why not.

**Comment by lukaseder** [comment]
Hmm, this seems to have shipped already in 3.14, but not closed. Well, I'll document the feature in the release notes of 3.15, then. Also, the `DefaultConverterProvider` (added in 3.14: https://github.com/jOOQ/jOOQ/issues/3896) does not yet know about this conversion.

**Comment by lukaseder** [comment]
Note that if `Converter` implementations aren't being attached at code generation time, there will be a type safe way to convert from any data type in jOOQ 3.15, that doesn't use reflection. If you have:

```kotlin
data class Race(
  val id: Long,
  val distance: RaceDistance
)
```

You can write a query like this (I'm using Java for the example):

```java
ctx.select(RACE.ID, RACE.DISTANCE.convertFrom(RaceDistance::new))
   .from(RACE)
   .fetch(mapping(Race::new))
```

The method names might still change until this is released, but you get the idea. Of course, I still think that attaching the `RaceDistance` type to the `RACE.DISTANCE` column at code generation time is the way to go here.
